chipflow_lib/platforms/_software.py

```diff
@@ -0,0 +1,118 @@
+# SPDX-License-Identifier: BSD-2-Clause
+
+import logging
+import warnings
+
+from collections import defaultdict
+
+from amaranth import Fragment
+from amaranth.hdl import UnusedElaboratable
+from amaranth.hdl._ir import Design
+from amaranth_soc import wishbone
+from amaranth_soc.wishbone.sram import WishboneSRAM
+from pydantic import TypeAdapter
+
+from .. import ChipFlowError
+from ._utils import get_software_builds
+from ._signatures import DRIVER_MODEL_SCHEMA, DriverModel, DATA_SCHEMA, SoftwareBuild
+from ._annotate import submodule_metadata
+from ..software.soft_gen import SoftwareGenerator
+
+
+logger = logging.getLogger(__name__)
+__all__ = []
+
+
+def get_windows(wb_decoder):
+    def _translate(subwindow, window, window_name, window_range):
+        # Accessing a resource through a dense and then a sparse window results in very strange
```

Comment:
it'd be nice to do something like `window_start, window_stop, window_ratio = window_range` and then use those variables here rather than everything being indexes that are hard to read

Reply:
good idea